Fix mpprintf argument type errors

jepler opened the pull request:
### Summary

I have developed a gcc plugin for checking mp_printf format strings vs argument types at compile time. https://github.com/micropython/micropython/pull/17556 is the fixes plus the checker, while this PR is only the fixes.

### Testing

I added tests to the coverage build for stuff I felt was not yet adequately covered.

jepler committed:
core: Cast type names to qstr explicitly.

The name field of type objects is of type uint16_t for efficiency,
but when the type is passed to mp_printf it must be cast explicitly
to type qstr.

These locations were found using an experimental gcc plugin
for mp_printf error checking, cross-building for x64 windows
on Linux.

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler committed:
core: Fix mp_printf integer size mismatches.

The type of the argument must match the format string. Add
casts to ensure that they do.

It's possible that casting from `size_t` to `unsigned` loses
the correct values by masking off upper bits, but it seems likely
that the quantities involved in practice are small enough that
the %u formatter (32 bits on most platforms, 16 on pic16bit) will
in fact hold the correct value.

The alternative, casting to a wider type, adds code size.

These locations were found using an experimental gcc plugin
for mp_printf error checking, cross-building for x64 windows
on Linux.

In one case there was already a cast, but it was written
incorrectly and did not have the intended effect.

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler committed:
coverage: Provide argmuents of expected types.

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler committed:
coverage: Cast type names to qstr explicitly.

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler committed:
coverage: Avoid type checking an invalid string.

we still want this not to crash a runtime but the
new static checker wouldn't like it.

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler committed:
coverage: Cast values to int for printing.

During the coverage test, all the values encountered are within the
range of %d.

These locations were found using an experimental gcc plugin
for mp_printf error checking.

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler committed:
examplemodule: Cast arguments to printf.

These locations were found using an experimental gcc plugin
for mp_printf error checking.

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler committed:
objcell: Fix printing cell ID.

On the nanbox build, `o->obj` is a 64-bit type but `%p` formats
a 32-bit type, leading to undefined behavior.

Print the cell's ID as a hex integer instead.

This location was found using an experimental gcc plugin for mp_printf
error checking.

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler committed:
modlwip: Print timeout with correct format string.

As timeout is of type `mp_int_t`, it must be printed with INT_FMT.

Before, the compiler plugin produced an error in the PYBD_SF6
build, which is a nanboxing build with 64-bit ints.

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler committed:
stm32: Add casts when printing small integers.

All these arguments are of type `mp_{u,}int_t`, but the actual
value is always a small integer. Cast it so that it can format
with the %d/%u formatter.

Before, the compiler plugin produced an error in the PYBD_SF6
build, which is a nanboxing build with 64-bit ints.

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler committed:
netutils: Cast the ticks value before printing.

Before, the compiler plugin produced an error in the PYBD_SF6
build, which is a nanboxing build with 64-bit ints.

I made the decision here to cast the value even though some
significant bits might be lost after 49.7 days. However, the
format used is "% 8d", which produces a consistent width
output for small ticks values (up to about 1.1 days). I judged
that it was more valuable to preserve the fixed width display
than to accurately represent long time periods.

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler committed:
coverage: Cast values to fit.

This fixes the following diagnostic produced by the plugin:

```
error: argument 3: Format ‘%x’ requires a ‘int’ or
    ‘unsigned int’ (32 bits), not ‘long unsigned int’ [size 64]
    [-Werror=format=]
```

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler committed:
coverage: Remove unused printf args.

Signed-off-by: Jeff Epler <jepler@gmail.com>

dpgeorge commented:
If `HEX_FMT` is really needed, maybe a simpler change would be to delete as many defines of `INT_FMT` and `UINT_FMT` as possible and let the defaults from `py/mpconfig.h` be used instead?

jepler committed:
mpprint: Fix printing pointers with upper bit set.

On a build like nanbox, mp_uint_t is wider than u/intptr_t.
Using a signed type for fetching pointer values resulted in
erroneous results: like `<function f at 0xfffffffff7a60bc0>`
instead of `<function f at 0xf7a60bc0>`.

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler commented:
Introducing and using HEX_FMT allows the nanbox variant to print cells correctly:
```
<closure <function <lambda> at 0xf79e7300> at f79e7320, n_closed=1 <cell f79e72a0 []> >
<closure <function <lambda> at 0xf79e7360> at f79e7380, n_closed=1 <cell 1000000000003 1> >
<closure <function <lambda> at 0xf79e73c0> at f79e73e0, n_closed=1 <cell c00d1eb851eb851f 3.14> >
<closure <function <lambda> at 0xf79e7440> at f79e7460, n_closed=1 <cell 200000000002b '__main__'> >
```
If the argument is cast to a pointer type and the `%p` format is used, then only the lower 32 bits are printed:
```
<closure <function <lambda> at 0xf79d8300> at f79d8320, n_closed=1 <cell f79d82a0 []> >
<closure <function <lambda> at 0xf79d8360> at f79d8380, n_closed=1 <cell 3 1> >
<closure <function <lambda> at 0xf79d83c0> at f79d83e0, n_closed=1 <cell 51eb851f 3.14> >
<closure <function <lambda> at 0xf79d8440> at f79d8460, n_closed=1 <cell 2b '__main__'> >
```

dpgeorge commented:
> I noticed I introduced a bug with printing pointers in #17618. I put a fix for it here but can break it out to its own PR if desired: [333665b](https://github.com/micropython/micropython/commit/333665bc140d494132c7b888a64a4e138a09c831)

It's OK to keep it here.

jepler committed:
ports: Eliminate {U,}INT_FMT where redundant.

The default definition in py/mpconfig.h is %u/%d, so
these can be removed.

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler committed:
various: Define HEX_FMT.

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler commented:
The one remaining failure is `1 tests failed: extmod/select_poll_eintr.py`

dpgeorge commented on py/objcell.c:
```diff
@@ -30,7 +30,8 @@
     (void)kind;
     mp_obj_cell_t *o = MP_OBJ_TO_PTR(o_in);
-    mp_printf(print, "<cell %p ", o->obj);
+    MP_STATIC_ASSERT(sizeof(mp_obj_t) == sizeof(mp_uint_t));
```
I don't think it makes sense to have this assert here, because this property is already fundamental to the entire runtime.

It's documented somewhat at the top of `py/obj.h`, where `mp_obj_t` is defined (although the comment there could be extended to note that `mp_uint_t` and `mp_int_t` need to be the same size as `mp_obj_t` (see also the definition of `MP_BYTES_PER_OBJ_WORD` in `py/mpconfig.h`).

If anything, this assert could go at the start of `mp_init()`.

dpgeorge replied:
Did you commit this change?

jepler commented on extmod/modlwip.c:
```diff
@@ -907,7 +907,7 @@ static const mp_obj_type_t lwip_socket_type;
 static void lwip_socket_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
     lwip_socket_obj_t *self = MP_OBJ_TO_PTR(self_in);
-    mp_printf(print, "<socket state=%d timeout=%d incoming=", self->state, self->timeout);
+    mp_printf(print, "<socket state=%d timeout=" INT_FMT " incoming=", self->state, self->timeout);
```
done

dpgeorge reviewed (approved):
Thanks for updating, looks good now.

dpgeorge commented:
Rebased and merged in 0c8d35b322554293c0847a618fa90760cefa1b60 through ab620f40844a837546497252849933ef2f685f4d
